dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java

```diff
@@ -738,114 +737,87 @@ private int startFetch()
         }
 
         //------------------------------------------------------------------------------
-        // Read the response length from the request
+        // Read the response from the request
         //------------------------------------------------------------------------------
 
-        int len = 0;
-        try
-        {
-            len = readReplyLen();
-        }
-        catch (HpccFileException e)
+        boolean inTokenRetry = false;
+        do
         {
-            prefetchException = e;
+            RowServiceResponse response = null;
             try
             {
-                close();
+                response = readResponse();
             }
-            catch(Exception ie){}
-            return -1;
-        }
-
-        // If len == 0 we have finished reading the file
-        if (len == 0)
-        {
-            try
+            catch (HpccFileException e)
             {
-                close();
+                prefetchException = e;
+                try
+                {
+                    close();
+                }
+                catch(Exception ie){}
+                return -1;
             }
-            catch (IOException e)
+
+            // If len < 0 we have finished reading the file
+            if (response.len < 0)
             {
-                prefetchException = new HpccFileException(e.getMessage());
+                try
+                {
+                    close();
+                }
+                catch (IOException e)
+                {
+                    prefetchException = new HpccFileException(e.getMessage());
+                }
+                return -1;
             }
-            return -1;
-        }
 
-        // Invalid response len, should be at least 4
-        if (len < 4)
-        {
-            prefetchException = new HpccFileException("Early data termination, no handle");
-            try
+            if (response.errorCode != RFCCodes.RFCStreamNoError)
             {
-                close();
+                prefetchException = new HpccFileException(response.errorMessage);
+                return -1;
             }
-            catch(Exception e){}
-            return -1;
-        }
 
-        //------------------------------------------------------------------------------
-        // Read the new handle, and retry with the token if it is invalid
-        //------------------------------------------------------------------------------
+            //------------------------------------------------------------------------------
+            // Retry with the token if handle is invalid
+            //------------------------------------------------------------------------------
 
-        try
-        {
-            this.handle = dis.readInt();
-            if (this.handle == 0)
+            this.handle = response.handle;
+            if (this.handle == 0 && !inTokenRetry)
```

Comment:
The handle at this point will always either be 0 or greater than 0, but it would probably be better to change that to this.handle <= 0; In case that changes in the future.